dataset_example.py

```diff
@@ -0,0 +1,162 @@
+import typing
+from abc import ABC
+
+import numpy as np
+import pandas as pd
+import pyarrow as pa
+
+from flytekit import kwtypes, task, workflow
+from flytekit.types.schema import SchemaFormat
+from flytekit.types.structured.structured_dataset import (
+    FLYTE_DATASET_TRANSFORMER,
+    DatasetDecodingHandler,
+    DatasetEncodingHandler,
+    FlyteDataset,
+    FlyteDatasetMetadata,
+)
+
+PANDAS_PATH = "/tmp/pandas.pq"
+NUMPY_PATH = "/tmp/numpy.pq"
+BQ_PATH = "bq://photo-313016:flyte.new_table3"
+
+# https://github.com/flyteorg/flyte/issues/523
+# We should also support map and list in schema column
+column_type = kwtypes(w=typing.Dict[str, typing.Dict[str, int]], x=typing.List[typing.List[int]], y=int, z=str)
+
+
+@task
+def t0() -> pd.DataFrame:
+    return pd.DataFrame({"Name": ["Tom", "Joseph"], "Age": [20, 22]})
+
+
+@task
+def t1(dataframe: pd.DataFrame) -> FlyteDataset[column_type]:
+    # S3 (parquet) -> Pandas -> S3 (parquet) default behaviour
+    return dataframe
+
+
+@task
+def t1a(dataframe: pd.DataFrame) -> FlyteDataset[column_type]:
+    # S3 (parquet) -> Pandas -> S3 (parquet) default behaviour
+    return FlyteDataset(dataframe=dataframe, remote_path=PANDAS_PATH)
+
+
+@task
+def t2(dataframe: pd.DataFrame) -> pd.DataFrame:
+    # S3 (parquet) -> Pandas -> S3 (parquet)
+    return dataframe
+
+
+@task
+def t3(dataset: FlyteDataset[column_type]) -> FlyteDataset[column_type]:
+    # s3 (parquet) -> pandas -> s3 (parquet)
+    print("Pandas dataframe")
+    print(dataset.open_as(pd.DataFrame))
+    return dataset
+
+
+@task
+def t3a(dataset: FlyteDataset[column_type]) -> FlyteDataset[column_type]:
+    # This task will not do anything - no uploading, no downloading
+    return dataset
+
+
+@task
+def t4(dataset: FlyteDataset[column_type]) -> pd.DataFrame:
+    # s3 (parquet) -> pandas -> s3 (parquet)
+    return dataset.open_as(pd.DataFrame)
+
+
+@task
+def t5(dataframe: pd.DataFrame) -> FlyteDataset[column_type]:
+    # s3 (parquet) -> pandas -> bq
+    return FlyteDataset(dataframe=dataframe, remote_path=BQ_PATH, file_format=SchemaFormat.BIGQUERY)
+
+
+@task
+def t6(dataset: FlyteDataset[column_type]) -> pd.DataFrame:
+    # bq -> pandas -> s3 (parquet)
+    df = dataset.open_as(pd.DataFrame)
+    return df
+
+
+@task
+def t7(df1: pd.DataFrame, df2: pd.DataFrame) -> (FlyteDataset[column_type], FlyteDataset[column_type]):
+    # df1: pandas -> bq
+    # df2: pandas -> s3 (parquet)
+    return FlyteDataset(dataframe=df1, remote_path=BQ_PATH, file_format=SchemaFormat.BIGQUERY), df2
```

Comment:
BIGQUERY is a storage location, but we still need to set a format for `StructuredDatasetMetadata.format`. 
The name need to be discussed